### PR TITLE
Refactor the preserve_host specs

### LINKED ISSUE #2052: Kong looping, with a 503
Chris Heald @cheald 17:42
Hey folks, I'm having some issues with 0.10.0rc3 - occassionally a request will 503 and go into an insane loop where it is re-requested dozens or hundreds of times and saturate nginx.
I'm trying to build openresty --with-debug so I can peek into the nginx internals and see what's going on, but doing so results in:

```
2017/02/07 16:41:21 [debug] /usr/local/openresty/bin/resty -V: 'nginx version: openresty/1.11.2.2'
2017/02/07 16:41:21 [debug] got ok: false, true, , nginx version: openresty/1.11.2.2
2017/02/07 16:41:21 [debug] OpenResty 'resty' executable not found at /usr/local/openresty/bin/resty
./kong/cmd/start.lua:34: ./kong/cmd/start.lua:20: could not find OpenResty 'resty' executable. Kong requires version 1.11.2.1 to 1.11.2.2
```

That "got ok" line is me printing the result of pl_utils.executeex(cmd) as found in prefix_handler.lua:is_openresty
The "success" flag is coming back negative, even though: $? on running that command manually is 0. Any ideas?
(Side note: I'm suspicious that the request storming is due to the internal DNS resolver - we're targeting a route53 domain which performs weighted resolution to a pair of AWS ELBs with this - the ELBs change IPs regularly and I've noticed that the 503 storms happen when the TTL on the DNS record is due to expire. But I need a debug build to confirm that.)

Chris Heald @cheald 18:35
http://i.imgur.com/JBWEp6i.png

is what the storming looks like, FWIW - I've confirmed that the origin isn't down when this happens, but Kong seems to lose the ability to resolve requests to the backend. I dropped Varnish in front of Kong in always-pass mode so that I can introspect traffic and I'm observing some really weird behavior where requests with the same Kong correlation ID seem to be replayed dozens of times per second. You can see when HTTP 200 traffic drops to nothing, everything starts erroring. The error traffic is a multiple of the "normal" traffic which suggests to me that there's some kind of weird amplification thing happening where failed requests are being replayed over and over. This happens even when I set retries=0 in the API config.

Chris Heald @cheald 19:20
compiling openresty with --without-luajit-lua52 fixes the startup issue. Now on to the debugging.

Thijs Schreijer @Tieske 19:55
@cheald the only place with 503's in the code is here; https://github.com/Mashape/kong/blob/64f5e9b05d25c5778cc6db927abe7b2f317969cc/kong/core/balancer.lua#L276-L308
But that indicates that there should at least be log messages.

Chris Heald @cheald 19:56
Yeah, I found those, I'm pretty sure that's not it. I think this is an actual 503 off the backend for some reason, but I haven't been able to identify why yet, and even in that case I wouldn't expect a storm of retries to choke out kong
I've got full nginx debug logs of the issue now, though, I'm trolling through those

Thijs Schreijer @Tieske 19:57
I cannot find it anymore, but in the last days there was someone reporting some sort of loopback issue, where Kong would proxy the request to itself, and go into a loop. Until the uri (which was extended on each request) became to large and an error was returned.

Chris Heald @cheald 19:58
That's the behavior I'm seeing, yes
The headers are becoming too large, rather than the URI itself, but that seems to be the general pattern

Thijs Schreijer @Tieske 20:03
iirc he had something like uris=/ and upstream_url=http://somehost/testand then requested http://kong/test which kept appending /test to the uri in a loop

Chris Heald @cheald 20:04
Right now Kong is just a straight pass-through for this config - upstream URL doesn't do any path modification

Thijs Schreijer @Tieske 20:04
can you share the config?
api config

Chris Heald @cheald 20:04
yeah, sec
```json
     {
      "uris": [
        "/"
      ],
      "id": "75a518ec-5bc7-43b3-b30e-f6351b6f73fc",
      "retries": 0,
      "preserve_host": true,
      "created_at": 1486180008000,
      "upstream_url": "https://mashable-com-688160256.us-east-1.elb.amazonaws.com",
      "https_only": true,
      "strip_uri": false,
      "http_if_terminated": true,
      "name": "api.mashable.com",
      "hosts": [
        "api.mashable.com"
      ]
    },
```

Kong basically just dynamically sets the upstream, then passes that back to nginx to do the actual work, correct?

```
2017/02/07 18:50:51 [debug] 26351#0: *5256 connect to 172.28.1.148:53, fd:75 #8003
2017/02/07 18:50:51 [debug] 26351#0: *5256 epoll add event: fd:75 op:1 ev:80002001
2017/02/07 18:50:51 [debug] 26351#0: *5256 http cleanup add: 00007F56C05AEEE8
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket connect: 0
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket network address given directly
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket resolve retval handler
2017/02/07 18:50:51 [debug] 26351#0: *5256 UDP socket 78
2017/02/07 18:50:51 [debug] 26351#0: *5256 connect to 172.28.1.195:53, fd:78 #8004
2017/02/07 18:50:51 [debug] 26351#0: *5256 epoll add event: fd:78 op:1 ev:80002001
2017/02/07 18:50:51 [debug] 26351#0: *5256 http cleanup add: 00007F56C05AEFA0
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket connect: 0
2017/02/07 18:50:51 [debug] 26351#0: from head of free queue, alloc semaphore: 00007F56CE31C2C8
2017/02/07 18:50:51 [debug] 26351#0: http lua semaphore new: 00007F56CE31C2C8, resources: 0
2017/02/07 18:50:51 [debug] 26351#0: *5256 send: fd:75 34 of 34
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket calling receive() method
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket read timeout: 2000
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket receive buffer size: 4096
2017/02/07 18:50:51 [debug] 26351#0: *5256 lua udp socket read data: waiting: 0
2017/02/07 18:50:51 [debug] 26351#0: *5256 recv: fd:75 -1 of 4096
2017/02/07 18:50:51 [debug] 26351#0: *5256 recv() not ready (11: Resource temporarily unavailable)
```

This looks like a likely culprit - DNS lookups are failing
Is there an easy way to turn on logging of the internal DNS handler?

Thijs Schreijer @Tieske 20:55
It uses 2 libraries, the basic one from OpenResty, and on top a more high-level one (by us).
@cheald can you test it with the dns record set with ttl=0 ?

Chris Heald @cheald 20:57
Hm, I'll see if I can
I'm using DNS weighting to siphon off a portion of live traffic to test with

Thijs Schreijer @Tieske 20:58
here's why; the resty dns lib has a set of retries and timeouts, now the lib on top uses locking such that only one dns query is performed for any given name at the same time. When the result returns, all waiting requests will get the same dns answer and continue.
Might be that the lock is causing this when a query fails
test with ttl=0: this is an exception for the lock. When it is 0 there is a query for every incoming request, no locking.

Chris Heald @cheald 21:00
That seems plausible. I haven't figured out how to cause the error cascade at will yet - right now I'm basically just letting a small amount of traffic hit the endpoint until it falls over. Usually breaks inside of 30 minutes though.
The theory is that with ttl=0, we shouldn't see it break?

Thijs Schreijer @Tieske 21:01
exactly. By definition ttl=0 means 1 query per request, no locks

Chris Heald @cheald 21:01
Actually, bah, no, I can't set ttl=0. ELB maintains its own DNS
60 second TTL, I don't know that I can change that

### PR DESCRIPTION
**Pull request by cheald**
Specs are now explicit about testing the issue surfaced in #2052 (preservation of upstream_host)

### Summary

The changes in 37e8ed30c68d592ddd54cd3ccf6b3ccffe453f1a added specs which incidentally cover the issue of the incorrect upstream_host being returned from the router when `preserve_host = true`. This just refactors those tests to be explicit about that test, and to assert a single case per test.

### Full changelog

* Refactored router preserve_host tests

### Issues resolved

* Fulfills the requirements to close #2052

### CODE REVIEW

File: spec/01-unit/12-router_spec.lua [hunk 1]
```diff
@@ -771,31 +771,52 @@ describe("Router", function()
...
+        local host = "discard.com"
 
-        local api, _, upstream_host, _, host_header = router.exec(_ngx)
-        assert.same(use_case_apis[2], api)
-        assert.equal("httpbin.org", upstream_host)
-        assert.is_nil(host_header)
+        it("does not change the target upstream", function()
+          local _ngx = mock_ngx("GET", "/", { ["host"] = host })
+
+          local api, _, upstream_host, _, host_header = router.exec(_ngx)
```
Comment (thibaultcha) [inline spec/01-unit/12-router_spec.lua]: The unused `host_header` is making the linting fail. I think we can remove the last two variables `_, host_header` for this test.

Comment (cheald) [inline spec/01-unit/12-router_spec.lua]: What's the correct way to run linting, for future reference?

Comment (thibaultcha) [inline spec/01-unit/12-router_spec.lua]: `make lint` as documented here: https://github.com/Mashape/kong#tests

**Comment by thibaultcha** [comment]
Thanks! Merged with modifications to the commit message.

**Pull request merged**